src/inspector/main_thread_interface.h

```diff
@@ -0,0 +1,137 @@
+  virtual ~Request() {}
+};
+
+template <typename Target, typename Arg>
+class SingleArgumentFunctionCall : public Request {
+ public:
+  using Fn = void (Target::element_type::*)(Arg);
+
+  SingleArgumentFunctionCall(Target target, Fn fn, Arg argument)
+                             : target_(target),
+                               fn_(fn),
+                               arg_(std::move(argument)) {}
+
+  void Call() override {
+    Apply(target_, fn_, std::move(arg_));
+  }
+
+ private:
+  template <typename Element>
+  void Apply(std::shared_ptr<Element> target, Fn fn, Arg arg) {
+    if (target != nullptr)
+      (target.get()->*fn)(std::move(arg));
+  }
+
+  template <typename Element>
+  void Apply(std::weak_ptr<Element> target, Fn fn, Arg arg) {
+    Apply(target.lock(), fn, arg);
+  }
+
+  Target target_;
+  Fn fn_;
+  Arg arg_;
+};
+
+std::unique_ptr<v8_inspector::StringBuffer> Utf8ToStringView(
+    const std::string& message);
+
+using MessageQueue = std::deque<std::unique_ptr<Request>>;
+class MainThreadInterface;
+
+class MainThreadHandle : public std::enable_shared_from_this<MainThreadHandle> {
+ public:
+  explicit MainThreadHandle(MainThreadInterface* main_thread)
+                            : main_thread_(main_thread) {}
+  ~MainThreadHandle() {
+    CHECK_NULL(main_thread_);  // main_thread_ should have called Reset
+  }
+  std::unique_ptr<InspectorSession> Connect(
+      std::unique_ptr<InspectorSessionDelegate> delegate,
+      bool prevent_shutdown);
+  void Post(std::unique_ptr<Request> request);
+  template <typename Target, typename Fn, typename Arg>
+  void Post(Target target, Fn fn, Arg argument) {
+    Post(std::unique_ptr<Request>(new SingleArgumentFunctionCall<Target, Arg>(
+        target, fn, std::move(argument))));
+  }
+
+
+  Agent* GetInspectorAgent();
+  std::unique_ptr<InspectorSessionDelegate> MakeThreadSafeDelegate(
+      std::unique_ptr<InspectorSessionDelegate> delegate);
+  bool Expired();
+
+ private:
+  void Reset();
+
+  MainThreadInterface* main_thread_;
+  Mutex block_lock_;
+  int next_session_id_ = 0;
+
+  friend class MainThreadInterface;
+};
+
+class MainThreadInterface {
+ public:
+  MainThreadInterface(Agent* agent, uv_loop_t*, v8::Isolate* isolate,
+                      v8::Platform* platform);
+  ~MainThreadInterface();
+
+  void DispatchMessages();
+  void Post(std::unique_ptr<Request> request);
+  bool WaitForFrontendEvent();
+  std::shared_ptr<MainThreadHandle> GetHandle();
+  Agent* inspector_agent() {
+    return agent_;
+  }
+
+ private:
+  using AsyncAndInterface = std::pair<uv_async_t, MainThreadInterface*>;
+
+  static void DispatchMessagesAsyncCallback(uv_async_t* async);
+  static void CloseAsync(AsyncAndInterface*);
+
+  MessageQueue requests_;
+  // This queue is to maintain the order of the messages for the cases
+  // when we reenter the DispatchMessages function.
+  MessageQueue dispatching_message_queue_;
+  bool dispatching_messages_ = false;
+  Mutex state_lock_;
```

Comment:
Can you add a comment indicating which members exactly are protected by this lock?

Reply:
Done. Also, renamed the lock.